aip/client-libraries/4222.md

```diff
@@ -13,7 +13,108 @@ routing.
 
 ## Guidance
 
-Code generators **should** look at URI-based variables declared in the
+Code generators **must** use the annotations to generate client libraries that,
+on a per-rpc basis, extract routing information from the request payload and
+add that information to the routing header.
+
+There are two annotations that specify how to extract routing information from
+the request payload: 
+* the [`google.api.routing`][routing] that specifies how to construct 
+routing headers explicitly 
+* the [`google.api.http`][http] annotation that may specify how to construct 
+routing headers implicitly. 
+For any given rpc, if the explicit routing headers annotation is present, the code
+generators **must** use it. If the explicit routing headers annotation is absent,
+the code generators **must** parse the [`google.api.http`][http] annotation to see
+if it specifies routing headers implicitly, and use that specification.
+
+## Explicit Routing Headers (`google.api.routing`)
+For an unary or server-streaming rpc the code generator **must** look at the routing
+parameters specified in the [`google.api.routing`][routing] annotation, if present.
+Any given routing parameter specifies a field name and a pattern with exactly one
+named resource ID path segment. For example:
+
+```proto
+rpc CreateTopic(CreateTopicRequest) {
+  option (google.api.routing) = {
+    routing_parameters {
+      field: "parent"
+      path_template: "{project=projects/*}/**"
+  }
+}
+```
+**Note:** An empty `google.api.routing` annotation is acceptable. It means that no
+routing headers should be generated for the rpc, when they otherwise would be
+e.g. implicitly from the `google.api.http` annotation.
+
+**Note:** It is acceptable to omit the pattern in the resource id segment, `{parent}`
+for example, is equivalent to `{parent=*}` and **must** be parsed.
```

Comment:
`{parent=*}` or `{parent=**}`? You've got `**` in the note below, but `*` here, when I'd expect them to be the same.

Reply:
no, if you specify `{parent}` it is equivalent to `{parent=*}`, whereas an omitted path template is a shortcut for "anything goes" which is `{<it's name>=**}`. Added an extra example to disambiguate.